File: apps/vscode/src/tools/shape/manipulation/enlargeShape.ts
/**
 * Enlarge Shape tool implementation.
 * Scales annotation shapes by a multiplicative factor relative to an origin point.
 *
 * Spec: shared/tools/shape/manipulation/enlarge-shape.1.0.md
 */

import type { DebriefFeature } from '@debrief/schemas';
import { isAnnotationFeature } from '@debrief/schemas';
import type { MCPToolDefinition } from '../../../types/tool';

export interface EnlargeShapeParams {
  scale_factor?: number;
  origin?: number[] | null;
}

export const toolDefinition: MCPToolDefinition = {
  name: 'enlarge-shape',
  description:
    'Scales annotation shapes by a multiplicative factor relative to an origin point. Uses linear interpolation of geographic coordinate differences.',
  inputSchema: {
    type: 'object',
    properties: {
      features: { type: 'array', items: { type: 'object' } },
      params: {
        type: 'object',
        properties: {
          scale_factor: {
            type: 'number',
            description: 'Multiplicative scaling factor (>1 enlarges, <1 shrinks, 1 is no-op)',
            default: 3.0,
          },
          origin: {
            type: 'array',
            description: 'Scaling origin as [longitude, latitude]. Default: geometric centroid.',
            items: { type: 'number' },
          },
        },
      },
    },
  },
  annotations: {
    'debrief:selectionRequirements': [
      { kind: 'CIRCLE', min: 1 },
      { kind: 'RECTANGLE', min: 1 },
      { kind: 'LINE', min: 1 },
      { kind: 'TEXT', min: 1 },
      { kind: 'VECTOR', min: 1 },
    ],
    'debrief:category': 'shape/manipulation',
    'debrief:version': '1.0.0',
    'debrief:outputKind': 'mutation/shape/scaled',
    'debrief:uiCategory': 'calc',
  },
};

const ANNOTATION_KINDS = new Set(['CIRCLE', 'RECTANGLE', 'LINE', 'TEXT', 'VECTOR']);

/**
 * Compute arithmetic mean of vertices as the geometric centroid.
 */
function computeCentroid(geometry: { type: string; coordinates: unknown }): number[] {
  const geoType = geometry.type;
  const coords = geometry.coordinates;

  if (geoType === 'Point') {
    return [...(coords as number[])];
  }

  let vertices: number[][];

  if (geoType === 'Polygon') {
    const ring = (coords as number[][][])[0]!;
    // Exclude closing vertex if first == last
    if (
      ring.length > 1 &&
      ring[0]![0] === ring[ring.length - 1]![0] &&
      ring[0]![1] === ring[ring.length - 1]![1]
    ) {
      vertices = ring.slice(0, -1);
    } else {
      vertices = ring;
    }
  } else if (geoType === 'LineString') {
    vertices = coords as number[][];
  } else {
    vertices = coords as number[][];
  }

  const n = vertices.length;
  if (n === 0) {
    return [0, 0];
  }

  let sumLon = 0;
  let sumLat = 0;
  for (const v of vertices) {
    sumLon += v[0]!;
    sumLat += v[1]!;
  }

  return [sumLon / n, sumLat / n];
}

/**
 * Scale a single [lon, lat] coordinate relative to an origin.
 */
function scaleCoordinate(
  coord: number[],
  origin: number[],
  scaleFactor: number,
): number[] {
  let newLon = origin[0]! + (coord[0]! - origin[0]!) * scaleFactor;
  let newLat = origin[1]! + (coord[1]! - origin[1]!) * scaleFactor;

  // Clamp latitude to [-90, 90]
  newLat = Math.max(-90, Math.min(90, newLat));

  // Normalise longitude to [-180, 180]
  newLon = ((newLon + 180) % 360) - 180;

  return [newLon, newLat];
}

function scaleCoordsList(
  coords: number[][],
  origin: number[],
  scaleFactor: number,
): number[][] {
  return coords.map((c) => scaleCoordinate(c, origin, scaleFactor));
}

export function execute(
  features: DebriefFeature[],
  params: EnlargeShapeParams,
): DebriefFeature[] {
  const scaleFactor = params.scale_factor ?? 3.0;
  const origin = params.origin ?? null;

  if (scaleFactor < 0) {
    throw new Error('scale_factor must be >= 0');
  }

  const modified: DebriefFeature[] = [];

  for (const feature of features) {
    if (!isAnnotationFeature(feature)) {
      continue;
    }

    const kind = feature.properties.kind;

    if (!ANNOTATION_KINDS.has(kind)) {
      continue;
    }

    const geometry = feature.geometry as { type: string; coordinates: unknown } | undefined;
    if (!geometry) {
      continue;
    }

    // Determine scaling origin
    const scalingOrigin = origin ?? computeCentroid(geometry);

    // Scale factor of 1.0 is a no-op
    if (scaleFactor === 1.0) {
      modified.push(feature);
      continue;
    }

    const coords = geometry.coordinates;

    if (kind === 'CIRCLE' || kind === 'RECTANGLE') {
      // Polygon: scale all rings
      const polyCoords = coords as number[][][];
      geometry.coordinates = polyCoords.map((ring) =>
        scaleCoordsList(ring, scalingOrigin, scaleFactor),
      );
      if (kind === 'CIRCLE') {
        const circleProps = feature.properties as { center?: number[] };
        if (circleProps.center !== undefined) {
          circleProps.center = scaleCoordinate(circleProps.center, scalingOrigin, scaleFactor);
        }
      }
    } else if (kind === 'LINE') {
      geometry.coordinates = scaleCoordsList(coords as number[][], scalingOrigin, scaleFactor);
    } else if (kind === 'TEXT') {
      geometry.coordinates = scaleCoordinate(coords as number[], scalingOrigin, scaleFactor);
    } else if (kind === 'VECTOR') {
      geometry.coordinates = scaleCoordsList(coords as number[][], scalingOrigin, scaleFactor);
      const vectorProps = feature.properties as { origin?: number[] };
      if (vectorProps.origin !== undefined) {
        vectorProps.origin = scaleCoordinate(vectorProps.origin, scalingOrigin, scaleFactor);
      }
    }

    modified.push(feature);
  }

  if (modified.length === 0) {
    throw new Error('No annotation features found in input');
  }

  return modified;
}

File: services/calc/debrief_calc/tools/shape/manipulation/enlarge_shape.py
"""Enlarge Shape tool — scale annotation features by a multiplicative factor."""

from __future__ import annotations

from typing import Any

from debrief_calc.models import (
    ContextType,
    GeoJSONFeatureDict,
    SelectionContext,
    ToolCategoryEnum,
    ToolParameter,
)
from debrief_calc.registry import tool

ANNOTATION_KINDS = {"CIRCLE", "RECTANGLE", "LINE", "TEXT", "VECTOR"}


def compute_centroid(geometry: GeoJSONFeatureDict) -> list[float]:
    """
    Compute arithmetic mean of vertices as the geometric centroid.

    For Polygon: uses exterior ring, excluding closing vertex.
    For LineString: uses all coordinates.
    For Point: returns the coordinate itself.

    Args:
        geometry: GeoJSON geometry dict with 'type' and 'coordinates'.

    Returns:
        [longitude, latitude] centroid coordinate.
    """
    geo_type = geometry.get("type")
    coords = geometry.get("coordinates")

    if coords is None:
        return [0.0, 0.0]

    if geo_type == "Point":
        return list(coords)

    if geo_type == "Polygon":
        ring = coords[0]
        # Exclude closing vertex if first == last
        vertices = ring[:-1] if len(ring) > 1 and ring[0] == ring[-1] else ring
    elif geo_type == "LineString":
        vertices = coords
    else:
        vertices = coords

    n = len(vertices)
    if n == 0:
        return [0.0, 0.0]

    sum_lon = sum(v[0] for v in vertices)
    sum_lat = sum(v[1] for v in vertices)
    return [sum_lon / n, sum_lat / n]


def scale_coordinate(coord: list[float], origin: list[float], scale_factor: float) -> list[float]:
    """
    Scale a single [lon, lat] coordinate relative to an origin.

    Args:
        coord: [longitude, latitude] to scale.
        origin: [longitude, latitude] scaling origin.
        scale_factor: Multiplicative scaling factor.

    Returns:
        Scaled [longitude, latitude].
    """
    new_lon = origin[0] + (coord[0] - origin[0]) * scale_factor
    new_lat = origin[1] + (coord[1] - origin[1]) * scale_factor

    # Clamp latitude to [-90, 90]
    new_lat = max(-90.0, min(90.0, new_lat))

    # Normalise longitude to [-180, 180]
    new_lon = ((new_lon + 180) % 360) - 180

    return [new_lon, new_lat]


def _scale_coords_list(
    coords: list[list[float]], origin: list[float], scale_factor: float
) -> list[list[float]]:
    """Scale a list of [lon, lat] coordinates relative to an origin."""
    return [scale_coordinate(c, origin, scale_factor) for c in coords]


@tool(
    name="enlarge-shape",
    description=(
        "Scales annotation shapes by a multiplicative factor relative to an origin point. "
        "Uses linear interpolation of geographic coordinate differences. "
        "Supports CIRCLE, RECTANGLE, LINE, TEXT, and VECTOR annotations."
    ),
    input_kinds=["CIRCLE", "RECTANGLE", "LINE", "TEXT", "VECTOR"],
    output_kind="mutation/shape/scaled",
    context_type=ContextType.MULTI,
    category=ToolCategoryEnum.calc,
    parameters=[
        ToolParameter(
            name="scale_factor",
            type="number",
            description="Multiplicative scaling factor (>1 enlarges, <1 shrinks, 1 is no-op)",
            required=False,
            default=3.0,
            choices=[0.25, 0.5, 1.5, 2.0, 3.0, 5.0],
        ),
        ToolParameter(
            name="origin",
            type="string",
            description=(
                "Scaling origin as [longitude, latitude] JSON array. "
                "Default: geometric centroid of each shape."
            ),
            required=False,
            default=None,
        ),
    ],
)
def enlarge_shape(context: SelectionContext, params: dict[str, Any]) -> list[GeoJSONFeatureDict]:
    """
    Scale annotation shapes by a multiplicative factor relative to an origin.

    Args:
        context: SelectionContext with one or more annotation features.
        params: Parameters dict with 'scale_factor' (default 3.0)
                and optional 'origin' ([lon, lat] or None for centroid).

    Returns:
        List of modified features with scaled coordinates.
    """
    scale_factor = float(params.get("scale_factor", 3.0))

    # Parse origin — may be a list or a JSON string
    raw_origin = params.get("origin")
    if isinstance(raw_origin, str):
        import json

        origin = json.loads(raw_origin)
    elif isinstance(raw_origin, list):
        origin = raw_origin
    else:
        origin = None

    if scale_factor < 0:
        raise ValueError("scale_factor must be >= 0")

    modified = []
    for feature in context.features:
        props = feature.get("properties", {})
        kind = props.get("kind")

        if kind not in ANNOTATION_KINDS:
            continue

        geometry = feature.get("geometry", {})

        # Determine scaling origin
        scaling_origin = origin if origin is not None else compute_centroid(geometry)

        # Scale factor of 1.0 is a no-op — return feature unchanged
        if scale_factor == 1.0:
            modified.append(feature)
            continue

        coords = geometry.get("coordinates")

        if kind in ("CIRCLE", "RECTANGLE"):
            # Polygon geometry: scale all rings
            geometry["coordinates"] = [
                _scale_coords_list(ring, scaling_origin, scale_factor) for ring in coords
            ]
            if kind == "CIRCLE" and "center" in props:
                props["center"] = scale_coordinate(props["center"], scaling_origin, scale_factor)

        elif kind == "LINE":
            # LineString geometry: scale all coordinates
            geometry["coordinates"] = _scale_coords_list(coords, scaling_origin, scale_factor)

        elif kind == "TEXT":
            # Point geometry: scale single coordinate
            geometry["coordinates"] = scale_coordinate(coords, scaling_origin, scale_factor)

        elif kind == "VECTOR":
            # LineString geometry: scale all coordinates
            geometry["coordinates"] = _scale_coords_list(coords, scaling_origin, scale_factor)
            # Update origin property; preserve range and bearing
            if "origin" in props:
                props["origin"] = scale_coordinate(props["origin"], scaling_origin, scale_factor)

        modified.append(feature)

    if not modified:
        raise ValueError("No annotation features found in input")

    return modified
